fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/AddDecodeNodeForDictStringRule.java

```diff
@@ -898,69 +877,163 @@ private static OptExpression generateDecodeOExpr(DecodeContext context, List<Opt
         return result;
     }
 
-    public static class CouldApplyDictOptimizeVisitor extends ScalarOperatorVisitor<Boolean, Void> {
+    // Check if an expression can be optimized using a dictionary
+    // 1. If the input column is only a dictionary column and there are no unsupported expressions in this expression,
+    // then it must be able to use dictionary optimization
+    // 2. If the input column is multi-column, and if there are expressions in the path of the dictionary column
+    // that can use dictionary optimization, then it is also able to use dictionary optimization
+    // eg:
+    // select if (x = 1, dict, y) from table; couldn't use dictionary optimize. If rewritten as a dictionary
+    // optimization is meaningless
+    //
+    // select if (dict = 1, x, y) from table; could use dictionary optimize.
+    // Because we can save the overhead of string filtering
+
+    public static class CouldApplyDictOptimizeContext {
+        // can use cardinality optimized dictionary columns
+        public Set<Integer> sids;
+        // whether is worth using dictionary optimization
+        boolean worthApplied = false;
+        //
+        boolean couldAppliedOperator = false;
+        // indicates the existence of expressions that do not support optimization using dictionaries
+        boolean hasUnsupportedOperator = false;
+
+        void reset() {
+            couldAppliedOperator = false;
```

Comment:
why not reset worthApplied?

Reply:
If any child of an expression is worth optimizing. then it must be worth optimizing